File: neoforge/src/main/java/com/kotori316/testutil/reporter/CombinedReporter.scala
package com.kotori316.testutil.reporter

import cats.implicits.catsSemigroupalForMonoid
import cats.{Invariant, Monoid, MonoidK}
import net.minecraft.gametest.framework.{GameTestInfo, TestReporter}

case class CombinedReporter(reporters: Seq[TestReporter]) extends TestReporter {
  override def onTestFailed(pTestInfo: GameTestInfo): Unit =
    reporters.foreach(_.onTestFailed(pTestInfo))

  override def onTestSuccess(pTestInfo: GameTestInfo): Unit =
    reporters.foreach(_.onTestSuccess(pTestInfo))

  override def finish(): Unit =
    reporters.foreach(_.finish())

  def append(reporter: TestReporter): CombinedReporter =
    copy(reporters = reporters :+ reporter)
}

object CombinedReporter {
  val empty: CombinedReporter = CombinedReporter(Seq.empty)
  implicit final val monoidReporter: Monoid[CombinedReporter] =
    Invariant[Monoid].imap(MonoidK[Seq].algebra[TestReporter])(CombinedReporter.apply)(_.reporters)
}

File: test-utility/common/src/main/java/com/kotori316/testutil/reporter/CombinedReporter.java
package com.kotori316.testutil.reporter;

import net.minecraft.gametest.framework.GameTestInfo;
import net.minecraft.gametest.framework.TestReporter;

import java.util.ArrayList;
import java.util.List;

public class CombinedReporter implements TestReporter {
    private final List<TestReporter> reporters;

    public CombinedReporter(List<TestReporter> reporters) {
        this.reporters = reporters;
    }

    public CombinedReporter(TestReporter... reporters) {
        this(List.of(reporters));
    }

    @Override
    public void onTestFailed(GameTestInfo gameTestInfo) {
        reporters.forEach(r -> r.onTestFailed(gameTestInfo));
    }

    @Override
    public void onTestSuccess(GameTestInfo gameTestInfo) {
        reporters.forEach(r -> r.onTestSuccess(gameTestInfo));

    }

    @Override
    public void finish() {
        TestReporter.super.finish();
        reporters.forEach(TestReporter::finish);
    }

    public CombinedReporter append(TestReporter reporter) {
        var copy = new ArrayList<>(reporters);
        copy.add(reporter);
        return new CombinedReporter(copy);
    }
}
